### PR TITLE
feat: remove headers filtering

### LINKED ISSUE #1262: Undici strips out set-cookie headers, even when "credentials: 'include'" is set
## Bug Description
Undici strips out `set-cookie` headers, which is expected per the `fetch` API spec. What is not expected though is that the `set-cookie` headers are stripped out in the presence of `{ credentials: 'include' }` in a `fetch` request. Here is the relevant portion of the `fetch` API spec:
https://fetch.spec.whatwg.org/#example-cors-with-credentials

### server.js
```
const express = require('express')
const app = express()

app.get('/', (req, res) => {
  res.set({
    abc: '123',
    'set-cookie': 'hello=world',
  });
  res.send('Hello World!');
})

app.listen(3333, () => {
  console.log(`Example app listening on port ${port}`)
})
```

### client.js
```
const { fetch } = require('undici');

async function get() {
  const headers = await fetch('http://localhost:3333/', { credentials: 'include' })
    .then(async res => {
      for await (const chunk of res.body) {}
      return res.headers
    });

  console.log(`abc: ${headers.get('abc')}`);
  console.log(`set-cookie: ${headers.get('set-cookie')}`);
}

get();
```

With the server script running, the client script returns:
```
abc: 123
set-cookie: null
```

## Expected Behavior

In the presence of `{ credentials: 'include' }` I would expect `set-cookie` headers to _not_ be stripped from the response. The expected output of the client script should be:
```
abc: 123
set-cookie: hello=world
```

## Environment

Mac OS 12.2.1
Node v16.13.1

### Additional context

There is a `TODO` in Undici's `fetch` code where this behavior should be implemented:
https://github.com/nodejs/undici/blob/main/lib/fetch/index.js#L1369-L1375

### LINKED ISSUE #1463: `cookie` no longer able to be set in the Headers API as of v5.2.0
## Bug Description

Upgrading Node from 18.1 to 18.2 we are no longer able to set the `cookie` header in the `Headers` API.

Seems this is a result of https://github.com/nodejs/undici/pull/1337

## Reproducible By

```js
const headers = new Headers();
headers.append('cookie', 'bar=baz; foo=bar');

fetch('https://httpbin.org/cookies', {
  headers,
})
  .then(res => res.json())
  .then(console.log);
```

Running the above code on Node 18.1:

```js
{ cookies: { bar: 'baz', foo: 'bar' } }
```

On Node 18.2 (and undidi 5.2):

```js
{ cookies: {} }
```

Curiously if I don't use the `Headers` API and set the `cookie` header directly it works:

```js
fetch('https://httpbin.org/cookies', {
  headers: {
    cookie: 'bar=baz; foo=bar',
  },
})
  .then(res => res.json())
  .then(console.log);

// { cookies: { bar: 'baz', foo: 'bar' } }
```

## Expected Behavior

Would expect to be able to still set the `cookie` header because we can't use `document.cookie` like you can use in browser environments.

## Environment

* OSX 11.6
* Node 18.1 and Node 18.2 using `nvm`

### PR DESCRIPTION
**Pull request by KhafraDev**
Fixes #1463 
Fixes #1262 

@ronag's checklist [here](https://github.com/nodejs/undici/issues/1262#issuecomment-1132013882) should be completed:

1. "A WinterCG issue" - https://github.com/wintercg/fetch/issues/6
2. "It's clear in the code that all related logic is outside of spec and a reference motivating it." - Areas of the code where the spec mentioned "filtered" or "safelisted" headers has been commented.
3. "Clearly added to the README."
4. "Investigate what exactly deno, node-fetch and cloudflare does so that we don't introduce another variant with slight differences." - 

- Cloudflare has a [`Headers.prototype.getAll`](https://developers.cloudflare.com/workers/runtime-apis/headers#differences) method (deprecated & removed from the spec) that only works with `set-cookie` headers.
- Deno [doesn't do this filtering at all](https://deno.land/manual/runtime/web_platform_apis#spec-deviations) (confirmed on my end, a Deno maintainer also mentions this in the wintercg issue).
- node-fetch has a [`Headers.prototype.raw`](https://github.com/node-fetch/node-fetch#extract-set-cookie-header) method that returns the raw headers.

Clearly, following Deno is the best idea and closest to the spec itself. We aren't adding anything in, but simply removing constraints that don't apply to a server environment.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -288,6 +288,14 @@ const headers = await fetch(url)
...
 ```
 
+##### Forbidden and Safelisted Header Names
+
+* https://fetch.spec.whatwg.org/#cors-safelisted-response-header-name
+* https://fetch.spec.whatwg.org/#forbidden-header-name
+* https://fetch.spec.whatwg.org/#forbidden-response-header-name
+
+The [Fetch Standard](https://fetch.spec.whatwg.org) requires implementations to exclude certain headers from requests and responses. Undici does not filter out these headers.
+
```
Comment (ronag) [inline README.md:298]: A link to the wintercg issue.

File: lib/fetch/request.js [hunk 2]
```diff
@@ -421,7 +421,7 @@ class Request {
       // list, append header’s name/header’s value to this’s headers.
       if (headers.constructor.name === 'Headers') {
         for (const [key, val] of headers[kHeadersList] || headers) {
-          this[kHeaders].append(key, val)
+          this[kState].headersList.append(key, val)
```
Comment (ronag) [inline lib/fetch/request.js:424]: Why this change?

Comment (KhafraDev) [reply lib/fetch/request.js:424]: Without it, doing something along the lines of 
```js
import { fetch, Headers } from 'undici'

const headers = new Headers([['cookie', 'value']])

await fetch('http://example.com', { headers })
```

wouldn't work. My best guess is removing the filtered headers but I'm not entirely sure. I believe I also made a similar change for the same reason.

File: README.md [hunk 3]
```diff
@@ -288,6 +288,15 @@ const headers = await fetch(url)
...
 ```
 
+##### Forbidden and Safelisted Header Names
+
+* https://fetch.spec.whatwg.org/#cors-safelisted-response-header-name
+* https://fetch.spec.whatwg.org/#forbidden-header-name
+* https://fetch.spec.whatwg.org/#forbidden-response-header-name
+* https://github.com/wintercg/fetch/issues/6
+
+The [Fetch Standard](https://fetch.spec.whatwg.org) requires implementations to exclude certain headers from requests and responses. Undici does not filter out these headers.
```
Comment (mcollina) [inline README.md:298]: Could you add a bit of an explanation on _why_ we do it?

Comment (KhafraDev) [reply README.md:298]: Not sure how to word it entirely...

"In browser environments, these headers are forbidden so the user agent remains in full control over them. In Undici, these constraints are removed to give more control to the user."? I took some of the wording directly from the spec without making it seem like this is insecure, but I'm not sure if it's enough.

Comment (KhafraDev) [reply README.md:298]: Also noticed that undici's filtering wasn't really spec compliant as it allowed `sec-` and `proxy-` headers anyways.

File: lib/fetch/request.js (hunk 2 above)
Comment (ronag) [reply lib/fetch/request.js:424]: I'm iffy about landing this before we understand what's going on here. Last time we were in This situation I found a rather serious bug.

Comment (KhafraDev) [reply lib/fetch/request.js:424]: I think it's mostly confusing with having multiple Headers/HeadersList for each Request and confusion with distinguishing one from another.

https://github.com/nodejs/undici/blob/0e6427467e5c2d604dca32b7cb01c0b4979d25de/lib/fetch/request.js#L421
The spec says to use "this's headers" and undici uses `this[kHeaders]`.

https://github.com/nodejs/undici/blob/0e6427467e5c2d604dca32b7cb01c0b4979d25de/lib/fetch/request.js#L427
Then a few lines down the spec says to also fill "this's Headers" and `this[kState].headersList` is being used.

Without looking too far into it, removing `this[kHeaders]` would probably be beneficial, and use `this[kState].headersList` (maybe change it to a Headers instance?) instead. Or remove one of them, so there's less confusion on which should be used?

Comment (KhafraDev) [reply lib/fetch/request.js:424]: With looking into it, we can set `this[kHeaders]` to a Headers class and set `request.headersList` to `this[kHeaders].headersList`?

Comment (KhafraDev) [reply lib/fetch/request.js:424]: Using `.set` and `.append` on the `Headers` instance doesn't seem to update its HeadersList. Updating the list itself, however, does work.

Comment (ronag) [reply lib/fetch/request.js:424]: I'm open for changes that are consistent and follow the wording of the spec.

Comment (KhafraDev) [reply lib/fetch/request.js:424]: Found a solution and was able to remove the weird hack that was added last time this happened.

File: lib/fetch/request.js [hunk 4]
```diff
@@ -406,26 +406,26 @@ class Request {
...
       // 2. If init["headers"] exists, then set headers to init["headers"].
       if (init.headers !== undefined) {
         headers = init.headers
       }
 
       // 3. Empty this’s headers’s header list.
-      this[kState].headersList = new HeadersList()
-      this[kHeaders][kHeadersList] = this[kState].headersList
+      this[kHeaders] = new Headers()
+      this[kState].headersList = this[kHeaders][kHeadersList]
```
Comment (ronag) [inline lib/fetch/request.js:418]: This doesn’t seem to be quite the same as the spec?

Comment (KhafraDev) [reply lib/fetch/request.js:418]: `Headers` is just a wrapper around `HeadersList` so emptying the headers list is as easy as creating a new instance of headers.

`this[kState]`'s headersList has to be the same as `this[kHeaders][kHeadersList]` so we need to re-assign the value after creating a new Headers instance.

Comment (ronag) [reply lib/fetch/request.js:418]: I’m a little worried that changing the reference might have unintended side effects. Why not just empty the list as the spec says?

Comment (ronag) [reply lib/fetch/request.js:418]: Spliec(0)

Comment (KhafraDev) [reply lib/fetch/request.js:418]: The list is a class with a symbol key that's a map but even calling `.clear()` on it or setting it to a new Map causes the tests to break. This is the only solution that works. I don't think changing the reference matters because this only happens in the constructor before we can even access the properties.

Comment (ronag) [reply lib/fetch/request.js:418]: Something is happening since the tests fail…

Comment (KhafraDev) [reply lib/fetch/request.js:418]: Tests pass when setting it to a new Headers instance (as it currently is) but everything else causes them to fail.

**Review by mcollina — APPROVED** [review]
lgtm

**Review by ronag — COMMENTED** [review]
I'm still concerned about the 2 places where this PR diverges from the spec. However, I'm going on vacation tomorrow and won't have time to dig into it myself. Merge it if you wish but I'd be happy if you could resolve this before merging.

File: lib/fetch/request.js [hunk 5]
```diff
@@ -383,10 +383,10 @@ class Request {
     // 30. Set this’s headers to a new Headers object with this’s relevant
     // Realm, whose header list is request’s header list and guard is
     // "request".
-    this[kHeaders] = new Headers()
+    this[kHeaders] = new Headers(request.headersList)
```
[the remark this answers is not in the document]
Comment (KhafraDev) [reply lib/fetch/request.js:386]: We set this's headers to a new Headers instance and then below we set this's headers headersList to this's request's headersList... Backwards, but it works and follows the spec.

File: lib/fetch/request.js [hunk 6]
```diff
@@ -406,26 +406,26 @@ class Request {
...
-      this[kState].headersList = new HeadersList()
-      this[kHeaders][kHeadersList] = this[kState].headersList
+      this[kHeaders] = new Headers()
+      this[kState].headersList = this[kHeaders][kHeadersList]
 
       // 4. If headers is a Headers object, then for each header in its header
       // list, append header’s name/header’s value to this’s headers.
       if (headers.constructor.name === 'Headers') {
-        for (const [key, val] of headers[kHeadersList] || headers) {
+        for (const [key, val] of headers) {
```
[the remark this answers is not in the document]
Comment (KhafraDev) [reply lib/fetch/request.js:423]: header's Symbol.iterator is a reference to a map's Symbol.iterator, seemed like a weird hack that was added last time this was an issue (since `headers[kHeadersList]` will always be set cuz it's a Headers instance).

File: lib/fetch/request.js (hunk 4 above)
Comment (KhafraDev) [reply lib/fetch/request.js:418]: Found this issue and I'll give you a hint about it, pretty funny...

look at the changes to Headers guard...

**Comment by KhafraDev** [comment]
@ronag Enjoy your vacation 😄, you were completely right - something was wrong with how the `guard` checks were being done now that the forbidden headers were no longer being checked as well (`append` and `set` were returning early so the headers weren't getting set). Only reason my changes worked was because `new Headers(...)` sets the guard to `none` by default, which doesn't have special behavior attached to it.

Everything follows the spec entirely now, and issues in the future should be easier now that I understand entirely what's happening!

**Comment by dsine-de** [comment]
Should this be available already in a current Node.js version? I don't see a `cookie` header being set on a Node.js fetch request when following a redirect with a `set-cookie` response header in Node.js v.18.8.0, like in the browser or e.g. in the Postman client.

**Comment by KhafraDev** [comment]
Yes, this is bundled in the current node version. If you make an issue or have a minimal reproduction I'd be glad to look into it.

**Comment by dsine-de** [comment]
Ok thank you, I don't know what's the decision if this should work with Node.js native fetch, but it seems Postman and the browser fetch have something like a cookie jar and if they get a `set-cookie` header in a request which redirects, they save that cookie and add it as a `cookie` header when following the redirect.

If this should work the same with Node.js fetch, I'll make a minimal repro example & issue.
